plugins/modules/system/kernel_blacklist.py

```diff
@@ -44,109 +46,64 @@
+        supports_check_mode=True,
+    )
 
-    def create_file(self):
-        if not self.checkmode and not os.path.exists(self.filename):
-            open(self.filename, 'a').close()
-            return True
-        elif self.checkmode and not os.path.exists(self.filename):
-            self.filename = os.devnull
-            return True
+    def __init_module__(self):
+        self.vars.filename = self.default_file if not self.vars.blacklist_file else self.vars.blacklist_file
+        self.vars.set('file_exists', os.path.exists(self.vars.filename), output=False, change=True)
+        if not self.vars.file_exists:
+            open(self.vars.filename, 'a').close()
+            self.vars.file_exists = True
+            self.vars.set('lines', [], change=True, diff=True)
         else:
-            return False
-
-    def get_pattern(self):
-        return r'^blacklist\s*' + self.module + '$'
-
-    def readlines(self):
-        f = open(self.filename, 'r')
-        lines = f.readlines()
-        f.close()
-        return lines
+            with open(self.vars.filename) as fd:
+                self.vars.set('lines', fd.readlines(), change=True, diff=True)
+        self.vars.set('is_blacklisted', self._is_module_blocked(), change=True)
 
-    def module_listed(self):
-        lines = self.readlines()
-        pattern = self.get_pattern()
+        self.pattern = re.compile(r'^blacklist\s+' + self.vars.name + '$')
 
-        for line in lines:
+    def _is_module_blocked(self):
+        for line in self.vars.lines:
             stripped = line.strip()
             if stripped.startswith('#'):
                 continue
-
-            if re.match(pattern, stripped):
+            if self.pattern.match(stripped):
                 return True
-
         return False
 
-    def remove_module(self):
-        lines = self.readlines()
-        pattern = self.get_pattern()
-
-        if self.checkmode:
-            f = open(os.devnull, 'w')
-        else:
-            f = open(self.filename, 'w')
-
-        for line in lines:
-            if not re.match(pattern, line.strip()):
-                f.write(line)
-
-        f.close()
-
-    def add_module(self):
-        if self.checkmode:
-            f = open(os.devnull, 'a')
-        else:
-            f = open(self.filename, 'a')
-
-        f.write('blacklist %s\n' % self.module)
-
-        f.close()
+    def state_absent(self):
+        if not self.vars.is_blacklisted:
+            return
+        self.vars.is_blacklisted = False
+        self.vars.lines = [line for line in self.vars.lines if not self.pattern.match(line.strip())]
+        if self.has_changed() and not self.module.check_module:
+            with open(self.vars.filename, 'w') as fd:
+                fd.writelines(self.vars.lines)
+
+    def state_present(self):
+        if self.vars.is_blacklisted:
+            return
+        self.vars.is_blacklisted = True
+        self.vars.lines = self.vars.lines + ['blacklist %s\n' % self.vars.name]
```

Comment:
If `self.vars.lines[-1]` does not end with a newline, the new last line will end up in the same line.

Reply:
Okie, I changed the state `present` logic to write the entire file out of `self.vars.lines` instead of appending, just like in the `absent` state. And since I was doing that, I moved that to `__quit_module__()` and made the logic a little bit more robust.